File: express-backend/src/services/provider/SteamGameSearchProvider.ts
import axios from 'axios';
import { config } from '../../config';
import { getDb } from '../../config/db';
import { GameSearchProvider } from '../provider/GameSearchProvider';
import {
  ExternalGameSearchResult,
  ProviderSearchResult,
  GameRecordSuggestion,
} from '../../dto/external-search';
import { RecordStatus } from '../../enums/RecordStatus';

function containsChinese(str: string): boolean {
  return /[一-鿿]/.test(str);
}

async function translateToEnglish(text: string): Promise<string | null> {
  try {
    const res = await axios.get('https://api.mymemory.translated.net/get', {
      params: { q: text, langpair: 'zh|en' },
      timeout: 5000,
    });
    const translated = res.data?.responseData?.translatedText;
    if (translated && translated.toLowerCase() !== text.toLowerCase()) {
      return translated;
    }
    return null;
  } catch {
    return null;
  }
}

// Steam 游戏搜索 Provider
// 使用 Steam Store Search API（结果按相关性排序，覆盖已下架游戏的重制版）
export class SteamGameSearchProvider implements GameSearchProvider {
  private static readonly PAGE_SIZE = 20;

  id(): string {
    return 'steam';
  }

  async search(query: string, page: number): Promise<ProviderSearchResult<ExternalGameSearchResult>> {
    const result: ProviderSearchResult<ExternalGameSearchResult> = {
      provider: this.id(),
      enabled: true,
      message: '',
      page,
      totalPages: 0,
      totalResults: 0,
      results: [],
    };

    if (!config.steam.apiKey) {
      result.enabled = false;
      result.message = '缺少 Steam Web API Key';
      return result;
    }

    if (!query) throw new Error('query must not be blank');
    const normalizedPage = Math.max(page, 1);

    // Steam 不支持中文搜索，翻译为英文
    let searchQuery = query.trim();
    if (containsChinese(searchQuery)) {
      const translated = await translateToEnglish(searchQuery);
      if (translated) {
        searchQuery = translated;
      }
    }

    try {
      const response = await axios.get('https://store.steampowered.com/api/storesearch/', {
        params: { term: searchQuery, l: 'english', cc: 'US' },
        timeout: 10000,
      });

      const items = response.data?.items ?? [];
      const total = response.data?.total ?? items.length;
      const totalPages = total === 0 ? 0 : Math.ceil(total / SteamGameSearchProvider.PAGE_SIZE);
      const fromIndex = Math.min((normalizedPage - 1) * SteamGameSearchProvider.PAGE_SIZE, items.length);
      const toIndex = Math.min(fromIndex + SteamGameSearchProvider.PAGE_SIZE, items.length);
      const pageItems = items.slice(fromIndex, toIndex);

      const steamAppIds = pageItems.map((i: any) => i.id).filter(Boolean);
      const existingMap = steamAppIds.length > 0
        ? await this.findExistingBySteamId(steamAppIds)
        : new Map<any, any>();

      const results: ExternalGameSearchResult[] = pageItems.map((item: any) => {
        const existing = existingMap.get(item.id) ?? null;
        const mapped: ExternalGameSearchResult = {
          provider: this.id(),
          rawgId: null,
          steamAppId: item.id ?? null,
          xboxId: null,
          psnId: null,
          title: item.name,
          posterUrl: `https://cdn.akamai.steamstatic.com/steam/apps/${item.id}/header.jpg`,
          releaseDate: null,
          overview: null,
          alreadyAdded: existing !== null,
          existingRecordId: existing?.id != null ? Number(existing.id) : null,
          suggestedRecord: null,
        };
        mapped.suggestedRecord = this.buildSuggestion(mapped);
        return mapped;
      });

      result.page = normalizedPage;
      result.totalPages = totalPages;
      result.totalResults = total;
      result.results = results;
    } catch {
      // 搜索失败返回空结果
    }

    return result;
  }

  private async findExistingBySteamId(ids: number[]): Promise<Map<any, any>> {
    const games = await getDb().game.findMany({ where: { steamAppId: { in: ids } } });
    return new Map(games.map((g) => [g.steamAppId!, g]));
  }

  private buildSuggestion(mapped: ExternalGameSearchResult): GameRecordSuggestion {
    return {
      rawgId: mapped.rawgId,
      steamAppId: mapped.steamAppId,
      xboxId: mapped.xboxId,
      psnId: mapped.psnId,
      title: mapped.title,
      posterUrl: mapped.posterUrl,
      status: RecordStatus.WANT,
      rating: null,
      shortReview: '',
    };
  }
}

File: legacy/java-backend/src/main/java/com/pixelreel/service/provider/SteamGameSearchProvider.java
package com.pixelreel.service.provider;

import com.pixelreel.config.SteamWebApiProperties;
import com.pixelreel.dto.external.ExternalGameSearchResult;
import com.pixelreel.dto.external.GameRecordSuggestion;
import com.pixelreel.dto.external.ProviderSearchResult;
import com.pixelreel.dto.steam.SteamAppListResponse;
import com.pixelreel.entity.Game;
import com.pixelreel.enums.RecordStatus;
import com.pixelreel.service.GameService;
import com.pixelreel.service.search.GameSearchProvider;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.client.RestClient;

@Service
@RequiredArgsConstructor
public class SteamGameSearchProvider implements GameSearchProvider {
  private final GameService gameService;

  @Qualifier("steamWebApiRestClient")
  private final RestClient steamWebApiRestClient;

  private final SteamWebApiProperties steamProperties;

  private final AtomicReference<List<SteamAppListResponse.SteamApp>> cachedApps = new AtomicReference<>();
  private volatile Instant lastRefresh = Instant.EPOCH;

  @Override
  public String id() {
    return "steam";
  }

  @Override
  public ProviderSearchResult<ExternalGameSearchResult> search(String query, int page) {
    ProviderSearchResult<ExternalGameSearchResult> result = new ProviderSearchResult<>();
    result.setProvider(id());

    if (!steamProperties.enabled()) {
      result.setEnabled(false);
      result.setMessage("Steam 未启用");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(steamProperties.apiKey())) {
      result.setEnabled(false);
      result.setMessage("缺少 Steam Web API Key");
      result.setPage(page);
      result.setTotalPages(0);
      result.setTotalResults(0);
      result.setResults(Collections.emptyList());
      return result;
    }

    if (!StringUtils.hasText(query)) {
      throw new IllegalArgumentException("query must not be blank");
    }

    int normalizedPage = Math.max(page, 1);
    List<SteamAppListResponse.SteamApp> apps = loadAppList();
    List<SteamAppListResponse.SteamApp> filtered = filterApps(apps, query);

    int pageSize = steamProperties.searchPageSize() == null ? 20 : steamProperties.searchPageSize();
    int total = filtered.size();
    int totalPages = total == 0 ? 0 : (int) Math.ceil(total / (double) pageSize);

    int fromIndex = Math.min((normalizedPage - 1) * pageSize, filtered.size());
    int toIndex = Math.min(fromIndex + pageSize, filtered.size());
    List<SteamAppListResponse.SteamApp> pageItems = filtered.subList(fromIndex, toIndex);

    Map<Long, Game> existingBySteamId = findExistingBySteamId(pageItems);

    List<ExternalGameSearchResult> results = new ArrayList<>();
    for (SteamAppListResponse.SteamApp item : pageItems) {
      ExternalGameSearchResult mapped = new ExternalGameSearchResult();
      mapped.setProvider(id());
      mapped.setSteamAppId(item.getAppId());
      mapped.setTitle(item.getName());
      mapped.setPosterUrl(null);

      Game existing = item.getAppId() == null ? null : existingBySteamId.get(item.getAppId());
      mapped.setAlreadyAdded(existing != null);
      mapped.setExistingRecordId(existing == null ? null : existing.getId());
      mapped.setSuggestedRecord(buildSuggestion(mapped));
      results.add(mapped);
    }

    result.setEnabled(true);
    result.setPage(normalizedPage);
    result.setTotalPages(totalPages);
    result.setTotalResults(total);
    result.setResults(results);
    return result;
  }

  private List<SteamAppListResponse.SteamApp> loadAppList() {
    int refreshMinutes = steamProperties.appListRefreshMinutes() == null ? 1440 : steamProperties.appListRefreshMinutes();
    boolean stale = Instant.now().isAfter(lastRefresh.plus(Duration.ofMinutes(refreshMinutes)));
    List<SteamAppListResponse.SteamApp> cached = cachedApps.get();
    if (!stale && cached != null) {
      return cached;
    }

    SteamAppListResponse response = steamWebApiRestClient
        .get()
        .uri(uriBuilder -> uriBuilder
            .path("/IStoreService/GetAppList/v1/")
            .queryParamIfPresent("key", Optional.ofNullable(steamProperties.apiKey()))
            .build())
        .retrieve()
        .body(SteamAppListResponse.class);

    List<SteamAppListResponse.SteamApp> apps = response == null || response.getResponse() == null
        ? Collections.emptyList()
        : response.getResponse().getApps();

    cachedApps.set(apps == null ? Collections.emptyList() : apps);
    lastRefresh = Instant.now();
    return cachedApps.get();
  }

  private List<SteamAppListResponse.SteamApp> filterApps(List<SteamAppListResponse.SteamApp> apps, String query) {
    if (apps == null) {
      return Collections.emptyList();
    }
    String needle = query.trim().toLowerCase(Locale.ROOT);
    return apps.stream()
        .filter(app -> StringUtils.hasText(app.getName()))
        .filter(app -> app.getName().toLowerCase(Locale.ROOT).contains(needle))
        .sorted(Comparator.comparing(SteamAppListResponse.SteamApp::getName))
        .toList();
  }

  private Map<Long, Game> findExistingBySteamId(List<SteamAppListResponse.SteamApp> items) {
    List<Long> ids = items.stream()
        .map(SteamAppListResponse.SteamApp::getAppId)
        .filter(value -> value != null)
        .distinct()
        .toList();
    if (ids.isEmpty()) {
      return Collections.emptyMap();
    }
    return gameService.lambdaQuery()
        .in(Game::getSteamAppId, ids)
        .list()
        .stream()
        .collect(Collectors.toMap(Game::getSteamAppId, Function.identity(), (a, b) -> a));
  }

  private GameRecordSuggestion buildSuggestion(ExternalGameSearchResult mapped) {
    GameRecordSuggestion suggestion = new GameRecordSuggestion();
    suggestion.setSteamAppId(mapped.getSteamAppId());
    suggestion.setTitle(mapped.getTitle());
    suggestion.setPosterUrl(mapped.getPosterUrl());
    suggestion.setStatus(RecordStatus.WANT);
    suggestion.setRating(null);
    suggestion.setShortReview("");
    return suggestion;
  }
}


